[C++20] [Modules] Don't generate the defintition for non-const available external variables

ChuanqiXu9 opened the pull request:
Close https://github.com/llvm/llvm-project/issues/93497

The root cause of the problem is, we mark the variable from other
modules as constnant in LLVM incorrectly. This patch fixes this problem
by not emitting the defintition for non-const available external
variables. Since the non const available externally variable is not
helpful to the optimization.

efriedma-quic reviewed (commented):
I think if a variable is GVA_AvailableExternally, and we can't emit a constant, we should just completely skip emitting the definition: there isn't any point to emitting an available_externally definition that doesn't actually contain any information the optimizer can use.

Not sure off the top of my head where that check belongs; might be okay to just stick it into EmitGlobalVarDefinition itself.

ChuanqiXu9 committed:
[C++20] [Modules] Don't generate the defintition for non-const available external variables

Close #93497

The root cause of the problem is, we mark the variable from other
modules as constnant in LLVM incorrectly. This patch fixes this problem
by not emitting the defintition for non-const available external
variables. Since the non const available externally variable is not
helpful to the optimization.

ChuanqiXu9 commented:
> I think if a variable is GVA_AvailableExternally, and we can't emit a constant, we should just completely skip emitting the definition: there isn't any point to emitting an available_externally definition that doesn't actually contain any information the optimizer can use.
> 
> Not sure off the top of my head where that check belongs; might be okay to just stick it into EmitGlobalVarDefinition itself.

Neat suggestion! I've applied it and the generated code looks better.

efriedma-quic commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -5341,6 +5341,15 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
+  // comment for ideas.
+  if (IsDefinitionAvailableExternally &&
+      (!D->hasConstantInitialization() ||
+       !D->getType().isConstantStorage(getContext(), true, true)))
```
Should we check for needsDestruction()?  (For example, consider a struct with a constexpr constructor, but a non-trivial destructor.)

ChuanqiXu9 replied:
(we think) the available externally variables don't need ctor or dtor in the current TU.  We mentioned this in the comment.

efriedma-quic replied:
No, I mean, please make sure we don't emit an available_externally definition in that case.

ChuanqiXu9 replied:
Got it. Makes sense. Done.

efriedma-quic reviewed (approved):
LGTM
